src/toil/lib/aws/s3.py

```diff
@@ -11,21 +11,484 @@
+        bucket.objects.all().delete()
+        bucket.object_versions.delete()
+        bucket.delete()
+        printq(f'\n * Deleted s3 bucket successfully: {bucket_name}\n\n', quiet)
+        logger.debug("Deleted s3 bucket successfully '%s'.", bucket_name)
+    except s3_client.exceptions.NoSuchBucket:
+        printq(f'\n * S3 bucket no longer exists: {bucket_name}\n\n', quiet)
+        logger.debug("S3 bucket no longer exists '%s'.", bucket_name)
+    except ClientError as e:
+        if get_error_status(e) != 404:
+            raise
+        printq(f'\n * S3 bucket no longer exists: {bucket_name}\n\n', quiet)
+        logger.debug("S3 bucket no longer exists '%s'.", bucket_name)
+
+
+@retry(errors=[BotoServerError])
+def bucket_exists(s3_resource, bucket: str) -> Union[bool, Bucket]:
+    s3_client = s3_resource.meta.client
+    try:
+        s3_client.head_bucket(Bucket=bucket)
+        return s3_resource.Bucket(bucket)
+    except (ClientError, s3_client.exceptions.NoSuchBucket) as e:
+        error_code = e.response.get('ResponseMetadata', {}).get('HTTPStatusCode')
+        if error_code == 404:
+            return False
+        else:
+            raise
+
+
+@retry(errors=[AWSServerErrors])
+def head_s3_object(bucket: str, key: str, header: Dict[str, Any], region: Optional[str] = None) -> HeadObjectOutputTypeDef:
+    """
+    Attempt to HEAD an s3 object and return its response.
+
+    :param bucket: AWS bucket name
+    :param key:  AWS Key name for the s3 object
+    :param header: Headers to include (mostly for encryption).
+        See: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/head_object.html
+    :param region: Region that we want to look for the bucket in
+    """
+    s3_client = session.client("s3", region_name=region)
+    return s3_client.head_object(Bucket=bucket, Key=key, **header)
+
+
 @retry(errors=[AWSServerErrors])
 def list_multipart_uploads(
     bucket: str, region: str, prefix: str, max_uploads: int = 1
 ) -> ListMultipartUploadsOutputTypeDef:
     s3_client = session.client("s3", region_name=region)
-    return s3_client.list_multipart_uploads(
-        Bucket=bucket, MaxUploads=max_uploads, Prefix=prefix
+    return s3_client.list_multipart_uploads(Bucket=bucket, MaxUploads=max_uploads, Prefix=prefix)
+
+
+@retry(errors=[BotoServerError])
+def copy_s3_to_s3(s3_resource, src_bucket, src_key, dst_bucket, dst_key, extra_args: Optional[dict] = None):
+    if not extra_args:
+        source = {'Bucket': src_bucket, 'Key': src_key}
+        # Note: this may have errors if using sse-c because of
+        # a bug with encryption using copy_object and copy (which uses copy_object for files <5GB):
+        # https://github.com/aws/aws-cli/issues/6012
+        # this will only happen if we attempt to copy a file previously encrypted with sse-c
+        # copying an unencrypted file and encrypting it as sse-c seems to work fine though
```

Comment:
The issue linked appears to apply only to the AWS CLI, and [not the API](https://github.com/aws/aws-cli/issues/6012#issuecomment-901416584), and also only describes adding encryption, not removing or changing it. Does it really not work if we do it through the API ourselves? (Or would we need tor each in and fiddle with headers manually?)